### PR TITLE
Add support for nested/grouped contests

### PR DESCRIPTION
**Pull request by nekodex**
Allows for the 'grouping' of contests to support showing multiple on the same page (e.g. for beatmap contests which require breakdown by difficulty/mode)

### CODE REVIEW

File: app/Http/Controllers/ContestsController.php [hunk 1]
```diff
@@ -36,16 +37,24 @@ public function index()
 
     public function show($id)
     {
-        $contest = Contest::with('entries', 'entries.contest', 'entries.user')->findOrFail($id);
+        $contest = Contest::findOrFail($id);
+
+        $ids = isset($contest->extra_options['children']) ? $contest->extra_options['children'] : [$id];
```
Comment (nanaya) [inline app/Http/Controllers/ContestsController.php]: $contest->extra_options['children'] ?? [$id]

File: resources/assets/coffee/react/contest-voting.coffee [hunk 2]
```diff
@@ -16,13 +16,14 @@
 *    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
 *
 ###
-propsFunction = ->
-  data = osu.parseJson('json-contest')
+propsFunction = (target) ->
+  srcNode = document.getElementById($(target).attr('data-src'))
```
Comment (nanaya) [inline resources/assets/coffee/react/contest-voting.coffee]: osu.parseJson target.dataset.src

File: resources/assets/less/bem/contest.less [hunk 3]
```diff
@@ -47,6 +48,55 @@
...
+    border-top-left-radius: 5px;
+    border-top-right-radius: 5px;
+
+    .white-link();
+    &:hover, &:visited, &:link, &:active {
+      .white-link();
+      text-decoration: none;
+    }
+
+    & .@{_top}__section-toggle {
```
Comment (nanaya) [inline resources/assets/less/bem/contest.less]: Doesn't need to be inside the `group-heading`?

```
&__section-toggle {
  transition...
  .@{_top}__group-heading[stuff=another-stuff] & {
    ...
  }

  .@{_top}__group-heading[stuff=stuff] & {
    ...
  }
}
```

File: resources/assets/less/bem/contest.less [hunk 4]
```diff
@@ -99,6 +149,10 @@
     }
   }
 
+  &__vote-star {
+    height: 34px;
```
Comment (nanaya) [inline resources/assets/less/bem/contest.less]: what is this for?

File: resources/views/contests/voting-multi.blade.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        <div class='contest__voting-notice'>{{trans('contest.voting.over')}}</div>
+    @endif
+    <div class='contest__accordian' id='contests-accordian'>
+        @foreach ($contests as $c)
+            <div class='panel contest__group{{ $loop->first ? ' panel-default' : '' }}'>
+                <a href="#{{$c->id}}" class='contest__group-heading' data-toggle='collapse' data-parent='#contests-accordian' aria-expanded='{{ $loop->first ? 'true' : 'false' }}'>
+                    <span>{!! $c->name !!}</span>
+                    <i class="contest__section-toggle fa fa-fw fa-chevron-down"></i>
+                </a>
+                <div class='contest__multi-panel collapse{{ $loop-> first ? ' in' : '' }}' id="{{$c->id}}">
```
Comment (nanaya) [inline resources/views/contests/voting-multi.blade.php]: cool syntax there (`$loop-> first`)

File: resources/views/contests/voting-multi.blade.php [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+--}}
+@extends('contests.base')
+
+@section('contest-content')
+    <div class="contest__description">{!! Markdown::convertToHtml($contest->description_voting) !!}</div>
+    @if ($contest->voting_ends_at !== null && $contest->voting_ends_at->isPast())
+        <div class='contest__voting-notice'>{{trans('contest.voting.over')}}</div>
+    @endif
+    <div class='contest__accordian' id='contests-accordian'>
+        @foreach ($contests as $c)
```
Comment (nanaya) [inline resources/views/contests/voting-multi.blade.php]:
```
@if (count($contests) === 1)
  @include('_contest')
@else
  @foreach (...)
  <collapse stuff...>
    @include('_contest')
  </>
@endif
```

File: resources/views/contests/voting-multi.blade.php [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                <div class='contest__multi-panel collapse{{ $loop-> first ? ' in' : '' }}' id="{{$c->id}}">
+                    @if ($c->type == 'art')
+                        <div class="js-react--contestArtList" data-src="contest-{{$contest->id}}"></div>
+                    @else
+                        <div class="js-react--contestList" data-src="contest-{{$contest->id}}"></div>
+                    @endif
+                    <script id="contest-{{$contest->id}}" type="application/json">
+                        {!! $c->defaultJson(Auth::user()) !!}
+                    </script>
+                    @if ($c->type == 'beatmap' && $c->extra_options['beatmapset_dl'])
```
Comment (nanaya) [inline resources/views/contests/voting-multi.blade.php]: needs an `isset` for the `extra_options`?

File: resources/assets/less/bem/tracklist.less [hunk 8]
```diff
@@ -45,6 +45,10 @@
     &--selected {
       background-color: rgba(141, 172, 182, 0.25);
     }
+    &:hover {
+      background-color: lighten(#293336, 10%);
+    }
+    transition: 0.1s;
```
Comment (nanaya) [inline resources/assets/less/bem/tracklist.less]: probably should be put before modifiers.

File: resources/assets/less/bem/contest.less [hunk 9]
```diff
@@ -47,6 +48,55 @@
...
+    background: #161a1b;
+    border: none;
+    .default-box-shadow();
+  }
+
+  &__group-heading {
+    display: flex;
+    justify-content: space-between;
+    padding: @spacing;
+    background: #161a1b;
```
Comment (nanaya) [inline resources/assets/less/bem/contest.less]: this and the radius doesn't seem to be needed

File: resources/views/contests/voting-multi.blade.php [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+@extends('contests.base')
+
+@section('contest-content')
+    <div class="contest__description">{!! Markdown::convertToHtml($contest->description_voting) !!}</div>
+    @if ($contest->voting_ends_at !== null && $contest->voting_ends_at->isPast())
+        <div class='contest__voting-notice'>{{trans('contest.voting.over')}}</div>
+    @endif
+    <div class='contest__accordian' id='contests-accordian'>
+        @foreach ($contests as $c)
+            <div class='panel contest__group{{ $loop->first ? ' panel-default' : '' }}'>
```
Comment (nanaya) [inline resources/views/contests/voting-multi.blade.php]: I tried removing the panel (and panel-default) class and it doesn't seem to change anything (apart of missing border radius)...

File: resources/assets/less/bem/contest.less [hunk 11]
```diff
@@ -47,6 +48,55 @@
...
+    border-top-left-radius: 5px;
+    border-top-right-radius: 5px;
+
+    .white-link();
+    &:hover, &:visited, &:link, &:active {
+      .white-link();
```
Comment (nanaya) [inline resources/assets/less/bem/contest.less]: the generated css seems... wrong. Probably just set the color directly or fix the `white-link` function.

File: resources/views/contests/voting-multi.blade.php (hunk 10 above)
Comment (nekodex) [inline resources/views/contests/voting-multi.blade.php]: bootstrap won't collapse the other sections (when opening a new one) without the panel class

File: resources/views/contests/voting-multi.blade.php (hunk 7 above)
Comment (nekodex) [inline resources/views/contests/voting-multi.blade.php]: If `extra_options` is null it doesn't error, but if it's `{}` it does...
i.e. `null['wut'] => null`, not error. TIL

File: resources/assets/coffee/react/contest-voting.coffee [hunk 12]
```diff
@@ -16,13 +16,13 @@
...
-propsFunction = ->
-  data = osu.parseJson('json-contest')
+propsFunction = (target) ->
+  data = osu.parseJson target.dataset.src
+
   return {
     contest: data.contest
     selected: data.userVotes
     options:
-      showDL: data.contest['type'] == 'beatmap'
```
Comment (nanaya) [inline resources/assets/coffee/react/contest-voting.coffee]: why is it removed?

File: resources/assets/less/bem/contest.less [hunk 13]
```diff
@@ -47,6 +48,52 @@
...
+    padding: @spacing;
+
+    .white-link();
+    &:hover, &:visited, &:link, &:active {
+      .white-link();
+      text-decoration: none;
+    }
+
+    &[aria-expanded='true'] {
+      .@{_top}__section-toggle {
```
Comment (nanaya) [inline resources/assets/less/bem/contest.less]: I think it's easier to see behavior of an element when it's grouped together where it's declared.

```
.parent {
  // stuff
  &[state] {
    .el {
      // stuff which may or may not override base .el
    }
  }
}

.el {
  // stuff
}
```

vs

```
.parent {
  // stuff
}

.el {
  // stuff
  .parent[some-state] & {
      // stuff which may or may not override base .el
  }
}
```

File: resources/assets/coffee/react/contest-voting.coffee (hunk 12 above)
Comment (nekodex) [inline resources/assets/coffee/react/contest-voting.coffee]: cuz no longer used~

**Pull request merged**